lib/ts/recipe/multifactorauth/recipe.ts

```diff
@@ -0,0 +1,508 @@
+        for (const func of this.getAllFactorsFromOtherRecipesFunc) {
+            const factorIdsRes = func(tenantConfig);
+            for (const factorId of factorIdsRes.firstFactorIds) {
+                if (!factorIds.includes(factorId)) {
+                    factorIds.push(factorId);
+                }
+            }
+        }
+        return factorIds;
+    };
+
+    addGetFactorsSetupForUserFromOtherRecipes = (func: GetFactorsSetupForUserFromOtherRecipesFunc) => {
+        this.getFactorsSetupForUserFromOtherRecipesFuncs.push(func);
+    };
+
+    checkForValidFirstFactor = async (tenantId: string, factorId: string, userContext: UserContext): Promise<void> => {
+        const tenantInfo = await Multitenancy.getTenant(tenantId, userContext);
+        if (tenantInfo === undefined) {
+            throw new SessionError({
+                type: SessionError.UNAUTHORISED,
+                message: "Tenant not found",
+            });
+        }
+        const { status: _, ...tenantConfig } = tenantInfo;
+
+        // we prioritise the firstFactors configured in tenant. If not present, we fallback to the recipe config
+        // if validFirstFactors is undefined, we assume it's valid
+        // Core already validates that the firstFactors are valid as per the logn methods enabled for that tenant,
+        // so we don't need to do additional checks here
+        let validFirstFactors = tenantConfig.firstFactors ?? this.config.firstFactors;
+
+        // We return a 401 if the factor is not a valid first factor. It's likely that the user
+        // is trying to setup a secondary factor without an active session.
+        // https://github.com/supertokens/supertokens-core/issues/554#issuecomment-1874931338
+        if (validFirstFactors !== undefined && !validFirstFactors.includes(factorId)) {
+            throw new SessionError({
+                type: SessionError.UNAUTHORISED,
+                message: "Session is required for secondary factors",
+                payload: {
+                    clearTokens: false,
+                },
+            });
+        }
+    };
+
+    checkIfFactorUserLinkedToSessionUser = (
+        sessionUser: User,
+        factorUser: User
+    ): { status: "OK" } | { status: "VALIDATION_ERROR"; reason: string } => {
+        // this is called during sign in operations to ensure the secondary factor user is linked to the session user
+        // we disallow factor setup by sign in, and is allowed only via sign up
+        if (sessionUser.id !== factorUser.id) {
+            return {
+                status: "VALIDATION_ERROR",
+                reason: "Cannot complete MFA because of security reasons. Please contact support. (ERR_CODE_009)",
+            };
+        }
+
+        return {
+            status: "OK",
+        };
+    };
+
+    checkAllowedToSetupFactorElseThrowInvalidClaimError = async (
+        tenantId: string,
+        session: SessionContainerInterface,
+        sessionUser: User,
```

Comment:
```suggestion
```

Reply:
This can be fetched in the function.. keep function signatures clean!

Reply:
refactored